src/Plugin/media/Source/Embeddable.php

```diff
@@ -0,0 +1,155 @@
+   *   A media item.
+   * @param string $attribute_name
+   *   Name of the attribute to fetch.
+   *
+   * @return mixed|null
+   *   Metadata attribute value or NULL if unavailable.
+   */
+  public function getMetadata(MediaInterface $media, $name) {
+    if ($this->hasUnstructured($media)) {
+      return $this->getUnstructuredMetadata($media, $name);
+    }
+    return parent::getMetadata($media, $name);
+  }
+
+  /**
+   * Gets the value for a metadata attribute for a given media item.
+   * This is an alternate version to account for unstructured embeds.
+   *
+   * @param \Drupal\media\MediaInterface $media
+   *   A media item.
+   * @param string $attribute_name
+   *   Name of the attribute to fetch.
+   *
+   * @return mixed|null
+   *   Metadata attribute value or NULL if unavailable.
+   */
+  public function getUnstructuredMetadata(MediaInterface $media, $name) {
+    switch ($name) {
+      case 'default_name':
+        if ($title = $this->getMetadata($media, 'title')) {
+          return $title;
+        }
+        elseif ($url = $this->getMetadata($media, 'url')) {
+          return $url;
+        }
+        return parent::getMetadata($media, 'default_name');
+
+      case 'thumbnail_uri':
+        parent::getMetadata($media, 'thumbnail_uri');
+      case 'type':
+      case 'title':
+      case 'author_name':
+      case 'author_url':
+      case 'provider_name':
+      case 'provider_url':
+      case 'cache_age':
+      case 'thumbnail_width':
+      case 'thumbnail_height':
+      case 'url':
+      case 'width':
+      case 'height':
+        return null;
+
+      case 'html':
+        return $media->get('field_media_embeddable_code')->getValue();
+
+      default:
+        break;
+    }
+    return NULL;
+
+  }
+
+
+  /**
+   * Is there a value for the oEmbed URL?
+   *
+   * @param \Drupal\media\MediaInterface $media
+   *   A media item.
+   *
+   * @return bool
+   *   TRUE means it has an Unstructured embed, FALSE means that field is empty
+   */
+  public function hasOEmbed(MediaInterface $media) {
+    if (!empty($media->get('field_media_embeddable_oembed')->getValue())) {
+      return TRUE;
+    }
+    return FALSE;
+  }
+
+  /**
+   * Is there a value for the Unstructured Embed?
+   *
+   * @param \Drupal\media\MediaInterface $media
+   *   A media item.
+   *
+   * @return bool
+   *   TRUE means it has an Unstructured embed, FALSE means that field is empty
+   */
+  public function hasUnstructured(MediaInterface $media) {
+    if (!empty($media->get('field_media_embeddable_code')->getValue())) {
+      return TRUE;
+    }
+    return FALSE;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getSourceFieldConstraints() {
+    return [
+      'embeddable' => [],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSourceFieldValue(MediaInterface $media) {
+    if ($this->hasUnstructured($media)) {
+      $source_field = 'field_media_embeddable_code';
```

Comment:
Some potentially helpful information.
https://www.drupal.org/project/drupal/issues/2934966

Reply:
Yeah, I override the `getSourceValue()` method because it has a `$media` passed in, and I can check to see if we're using the oEmbed field or the unstructured field.  So this always returns the right value, no matter which kind of an embed it is.

I wish I could override the `getSourceFieldDefinition()` function as well, but that gets called *without* a media element to test, so there's no way to know there what the source field is supposed to be.